pkg/inventory/inventory-client.go

```diff
@@ -24,455 +24,420 @@ import (
 		if err != nil {
 			return nil, err
 		}
-		identifiers[invName] = append(identifiers[invName], wrappedInvObjSlice...)
+		inventories = append(inventories, uInv)
 	}
-
-	return identifiers, nil
+	return inventories, nil
 }
 
-// deleteInventoryObjByName deletes the passed inventory object from the APIServer, or
-// an error if one occurs.
-func (cic *ClusterClient) deleteInventoryObjByName(ctx context.Context, obj *unstructured.Unstructured, dryRun common.DryRunStrategy) error {
-	if obj == nil {
-		return fmt.Errorf("attempting delete a nil inventory object")
-	}
-	if dryRun.ClientOrServerDryRun() {
-		klog.V(4).Infof("dry-run delete inventory object: not deleted")
+// CreateOrUpdate the in-cluster inventory
+// Updates the unstructured object, or creates it if it doesn't exist
+func (cic *UnstructuredClient) CreateOrUpdate(ctx context.Context, inv Inventory, opts UpdateOptions) error {
+	ui, ok := inv.(*UnstructuredInventory)
+	if !ok {
+		return fmt.Errorf("expected UnstructuredInventory")
+	}
+	if ui == nil {
+		return fmt.Errorf("inventory is nil")
+	}
+	// Attempt to retry on a resource conflict error to avoid needing to retry the
+	// entire Apply/Destroy when there's a transient conflict.
+	attempt := 0
+	var obj *unstructured.Unstructured
+	err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		defer func() {
+			attempt++
+		}()
+		create := false
+		var err error
+		obj, err = cic.client.Namespace(ui.GetNamespace()).Get(ctx, ui.GetName(), metav1.GetOptions{})
+		if apierrors.IsNotFound(err) {
+			create = true
+			obj = cic.newUnstructuredObject(&ui.SingleObjectInfo)
+		} else if err != nil {
+			return err
+		}
+		uObj, err := cic.toUnstructured(obj.DeepCopy(), ui)
+		if err != nil {
+			return err
+		}
+		if create {
+			klog.V(4).Infof("[attempt %d] creating inventory object %s/%s/%s",
+				attempt, cic.gvk, uObj.GetNamespace(), uObj.GetName())
+			obj, err = cic.client.Namespace(uObj.GetNamespace()).Create(ctx, uObj, metav1.CreateOptions{})
+			if err != nil {
+				return err
+			}
+		} else {
+			klog.V(4).Infof("[attempt %d] updating inventory object %s/%s/%s",
+				attempt, cic.gvk, uObj.GetNamespace(), uObj.GetName())
+			obj, err = cic.client.Namespace(uObj.GetNamespace()).Update(ctx, uObj, metav1.UpdateOptions{})
+			if err != nil {
+				return err
+			}
+		}
 		return nil
-	}
-
-	mapping, err := cic.getMapping(obj)
+	})
 	if err != nil {
 		return err
 	}
-
-	klog.V(4).Infof("deleting inventory object: %s/%s", obj.GetNamespace(), obj.GetName())
-	return cic.dc.Resource(mapping.Resource).Namespace(obj.GetNamespace()).
-		Delete(ctx, obj.GetName(), metav1.DeleteOptions{})
-}
-
-// getMapping returns the RESTMapping for the provided resource.
-func (cic *ClusterClient) getMapping(obj *unstructured.Unstructured) (*meta.RESTMapping, error) {
-	return cic.mapper.RESTMapping(obj.GroupVersionKind().GroupKind(), obj.GroupVersionKind().Version)
+	if cic.statusPolicy != StatusPolicyAll || opts.SkipStatusUpdate {
```

Comment:
hmm. this is curious. do we need both the struct field and the method options?

Reply:
Yes, they serves a different purposes.
- statusPolicy is configured when the inventory client is instantiated and completely enables/disables status updates
- `opts.SkipStatusUpdate` is set by the applier during inv_add_task to omit the status update during that task, because the object statuses are stale during that step. The statuses are updated with `inv_set_task`

Reply:
If InvAddTask doesn't update the status, it won't unblock the ResourceGroup controller. It needs to at least update the observedGeneration. That's how I fixed the race condition between controllers.

There's a bug we need to fix so that the inventory object status contains the correct status and conditions from kstatus, but we can fix that later. 

Even without that, the client still needs to update the inventory object status to set the strategy, actuation, and reconcile status. Setting the strategy declares the intent to apply or delete the object. Actuation & reconcile both get set to Pending because there's a new source that may not match the old source.

Reply:
> Even without that, the client still needs to update the inventory object status to set the strategy, actuation, and reconcile status. Setting the strategy declares the intent to apply or delete the object. Actuation & reconcile both get set to Pending because there's a new source that may not match the old source.

I was under the impression that the status would be stale during InvAddTask because InvAddTask doesn't update any of the object statuses. However I realized its buried in the old inventory client here, which sets the status for apply/prune objects: https://github.com/kubernetes-sigs/cli-utils/blob/master/pkg/inventory/inventory-client.go#L143. Given this realization, I think I'll push this logic up into the InvAddTask to update the object statuses before calling Update